### PR TITLE
Improvement of the SPIN package by removing a limitation

### PR DESCRIPTION
**Pull request by julient31**
make html**Summary**

This PR:
- corrected a limitation of the SPIN package: only one fix precession/spin or langevin/spin could be declared for each simulation. This is no longer the case.
- implemented a spin-transfer torque in fix precession/spin
- improved the test_problem setup

**Related Issue(s)**

None.

**Author(s)**

Julien Tranchida (Sandia National Labs), julien.tranchida1@gmail.com
Theophile Chirac (CEA, helped with the implementation of the spin-transfer torque), theophile.chirac@cea.fr

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Yes: all former input scripts will still be fully valid.

**Implementation Notes**

All validation tests of the SPIN package were successfully passed. This is important as changes were made in fix nve/spin, langevin/spin and precession/spin. 
The documentation was modified to account for the new possible interaction (SST in fox precession/spin)and successfully recompiled (make html). 

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

### CODE REVIEW

File: src/SPIN/fix_langevin_spin.cpp [hunk 1]
```diff
@@ -34,6 +35,8 @@
 #include "random_mars.h"
 #include "respa.h"
 #include "update.h"
+#include "compute.h"
```
Comment (akohlmey) [inline src/SPIN/fix_langevin_spin.cpp:38]: I don't see where compute.h would need to be included here.

File: src/SPIN/fix_precession_spin.cpp [hunk 2]
```diff
@@ -123,6 +148,13 @@ FixPrecessionSpin::FixPrecessionSpin(LAMMPS *lmp, int narg, char **arg) :
     nhz *= inorm;
   }
 
+  if (stt_flag) {
+    inorm = 1.0/sqrt(nsttx*nsttx + nstty*nstty + nsttz*nsttz);
```
Comment (akohlmey) [inline src/SPIN/fix_precession_spin.cpp:152]: There doesn't seem to be a check whether nsttx, netty and nsttz are valid input. from the context it seems they should be > 1. They most certainly must not be all 0 or else there would be a division by zero error here.

File: src/SPIN/fix_precession_spin.cpp [hunk 3]
```diff
@@ -144,6 +176,27 @@ FixPrecessionSpin::FixPrecessionSpin(LAMMPS *lmp, int narg, char **arg) :
     nc3y *= inorm;
     nc3z *= inorm;
   }
+  
+  if (hexaniso_flag) {
+    inorm = 1.0/sqrt(n6x*n6x + n6y*n6y + n6z*n6z);
```
Comment (akohlmey) [inline src/SPIN/fix_precession_spin.cpp:181]: same issue about division by zero from before.

File: src/SPIN/fix_precession_spin.cpp [hunk 4]
```diff
@@ -185,8 +238,9 @@ void FixPrecessionSpin::init()
   Kah = Ka/hbar;
   k1ch = k1c/hbar;
   k2ch = k2c/hbar;
+  K6h = K6/hbar;
 
-  if (utils::strmatch(update->integrate_style,"^respa")) {
```
Comment (akohlmey) [inline src/SPIN/fix_precession_spin.cpp:189]: this change should be reverted.

File: src/SPIN/fix_precession_spin.cpp [hunk 5]
```diff
@@ -225,7 +270,7 @@ void FixPrecessionSpin::init()
 
 void FixPrecessionSpin::setup(int vflag)
 {
-  if (utils::strmatch(update->integrate_style,"^verlet"))
+  if (strstr(update->integrate_style,"verlet"))
```
Comment (akohlmey) [inline src/SPIN/fix_precession_spin.cpp:273]: this change should be reverted

**Review by akohlmey — CHANGES_REQUESTED** [review]
@julient31 I have some suggestion for cosmetic changes, a couple of cases where upstream changes were accidentally reverted, and flagged some possible division by zero issues that should be checked for after parsing the input arguments.

**Comment by julient31** [comment]
Hey @akohlmey and @sjplimp 

Thanks so much for the reviews! I think I corrected everything. Let me know if you need me to modify something else!

Best,
Julien.

**Review by akohlmey — APPROVED** [review]

**Pull request merged**